File: lint-arwaky/archive/taxonomy/config_provider_error.py
"""config_provider_error — Configuration domain error types."""

from pydantic import BaseModel, ConfigDict, Field

from .config_identifier_vo import ConfigKey
from .file_path_vo import FilePath
from .error_value_vo import ErrorMessage, ExpectedValue, ActualValue


class ConfigError(BaseModel):
    """Invalid or missing configuration."""

    model_config = ConfigDict(frozen=True)

    key: ConfigKey
    message: ErrorMessage
    expected: ExpectedValue = Field(default_factory=lambda: ExpectedValue(value=""))
    actual: ActualValue = Field(default_factory=lambda: ActualValue(value=""))
    config_file: FilePath | None = None

    def __str__(self):
        file_info = f" in {self.config_file}" if self.config_file else ""
        return f"Config error on '{self.key}'{file_info}: {self.message}"

File: lint-arwaky/src-rust/taxonomy/config_provider_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ConfigError {
    pub key: ConfigKey,
    pub message: ErrorMessage,
    #[serde(default)]
    pub expected: Option<ExpectedValue>,
    #[serde(default)]
    pub actual: Option<ActualValue>,
    #[serde(default)]
    pub config_file: Option<FilePath>,
}

impl ConfigError {
    pub fn new(key: ConfigKey, message: ErrorMessage) -> Self {
        Self { key, message, expected: None, actual: None, config_file: None }
    }
}

impl std::fmt::Display for ConfigError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let file_info = self.config_file.as_ref().map(|p| format!(" in {}", p)).unwrap_or_default();
        write!(f, "Config error on '{}'{}: {}", self.key, file_info, self.message)
    }
}
